zqd/ingest/logreader.go

```diff
@@ -0,0 +1,124 @@
+package ingest
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"sync/atomic"
+
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio"
+	"github.com/brimsec/zq/zio/detector"
+	"github.com/brimsec/zq/zio/ndjsonio"
+	"github.com/brimsec/zq/zio/zngio"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/brimsec/zq/zqe"
+)
+
+type MultipartLogReader struct {
+	mr       *multipart.Reader
+	opts     zio.ReaderOpts
+	stopErr  bool
+	warnings []string
+	zreader  *zbuf.File
+	zctx     *resolver.Context
+	nread    int64
+	done     bool
+}
+
+func NewMultipartLogReader(mr *multipart.Reader, zctx *resolver.Context) *MultipartLogReader {
+	return &MultipartLogReader{
+		mr:   mr,
+		opts: zio.ReaderOpts{Zng: zngio.ReaderOpts{Validate: true}},
+		zctx: zctx,
+	}
+}
+
+func (m *MultipartLogReader) SetStopOnError() {
+	m.stopErr = true
+}
+
+func (m *MultipartLogReader) Read() (*zng.Record, error) {
+read:
+	if m.done {
+		return nil, nil
+	}
+	if m.zreader == nil {
+		zr, err := m.next()
+		if err == io.EOF {
+			m.done = true
+			return nil, nil
+		}
+		if zr == nil || err != nil {
+			return nil, err
+		}
+		m.zreader = zr
+	}
+	rec, err := m.zreader.Read()
+	if err != nil || rec == nil {
+		zr := m.zreader
+		m.zreader.Close()
+		m.zreader = nil
+		if err != nil {
+			if m.stopErr {
+				return nil, err
+			}
+			m.appendWarning(zr, err)
+		}
+		goto read
+	}
+	return rec, err
+}
+
+func (m *MultipartLogReader) next() (*zbuf.File, error) {
+next:
+	part, err := m.mr.NextPart()
+	if err != nil {
+		return nil, err
+	}
+	if part.FormName() == "json_type_config" {
```

Comment:
The zapi client sends `json_type_config` first, then files, which seems like what we'd want every client to do. How about checking that here, e.g. if a `json_type_config` is sent after logs, that is an error?

Reply:
It seems like maybe it could be a feature? For instance if you have two separate typing configs for two different sets of json logs. No one will ever do this, but it is there. Perhaps @philrz is crazy enough to try.

Reply:
Feature or opportunity for confusion is the question 🙂 ... we can leave it as is if you think this is useful.

Reply:
There definitely will be a future where different JSON sources with different typing needs will want to be ingested, though I see it as entirely possible we'll overhaul the approach to JSON typing at the same time we make that a first class operation in the app/tooling.